src/shamir.rs

```diff
@@ -0,0 +1,364 @@
+pub struct Share<F> {
+    x: u8,
+    y: F,
+}
+
+#[derive(Debug, Error)]
+pub enum ShareError {
+    #[error("Can't perform operation on shares as they don't have the same X coordinate: ({lhs} != {rhs})")]
+    MisalignedShares { lhs: u8, rhs: u8 },
+}
+
+impl SecretSharing {
+    /// Constructs a new instance, returning `ShamirError` if input values do not allow
+    /// to construct a valid Shamir secret sharing scheme.
+    ///
+    /// # Errors
+    /// Rejects the following inputs:
+    /// * k == 1
+    /// * k > n
+    pub fn new(k: NonZeroU8, n: NonZeroU8) -> Result<Self, Error> {
+        if k.get() == 1 || k > n {
+            Err(Error::BadSharingScheme {
+                k: k.get(),
+                n: n.get(),
+            })
+        } else {
+            Ok(Self {
+                k: k.get(),
+                n: n.get(),
+            })
+        }
+    }
+
+    /// Minimum number of shares required to reconstruct a secret
+    #[must_use]
+    pub fn threshold(&self) -> u8 {
+        self.k
+    }
+
+    /// Consumes and splits a given secret into `n` shares. only `k` <= `n` shares is required
+    /// to reconstruct it later
+    pub fn split<R: RngCore, F: PrimeField>(&self, secret: F, rng: R) -> Vec<Share<F>> {
+        // generate polynomial of k-1 degree
+        let coefficients = gen_polynomial::<R, F>(rng)
+            .take(self.k as usize)
+            .collect::<Vec<_>>();
+
+        let mut shares = Vec::with_capacity(self.n as usize);
+
+        // sample n points by evaluating the polynomial
+        for x in 1..=self.n {
+            let x_i = F::from(u64::from(x));
+            let mut y = secret;
+            for c in &coefficients {
+                y += (*c) * x_i;
+            }
+
+            shares.push(Share { x, y });
+        }
+
+        shares
+    }
+
+    /// Reconstructs a secret from a set of size at least k of shares.
+    ///
+    /// # Errors
+    /// Returns an error if there is no enough shares to reconstruct the secret or
+    /// if an error occurred while evaluating the polynomial
+    pub fn reconstruct<F: PrimeField>(&self, shares: &[Share<F>]) -> Result<F, Error> {
+        if shares.len() < self.threshold() as usize {
+            // SAFETY: len() fits into u8
+            #[allow(clippy::cast_possible_truncation)]
+            return Err(Error::NotEnoughShares {
+                required: self.threshold(),
+                actual: shares.len() as u8,
+            });
+        }
+
+        let mut r = F::zero();
+
+        for share_i in shares {
+            let mut x = F::one();
+            let mut denom = F::one();
+            for share_j in shares {
+                if share_j.x != share_i.x {
+                    x *= F::from(u64::from(share_j.x));
+                    denom *= F::from(u64::from(share_j.x)) - F::from(u64::from(share_i.x));
+                }
+            }
+
+            let maybe_denom = denom.invert();
+            if maybe_denom.is_none().into() {
```

Comment:
`CtOption` is a little clunky, in particular it is not an enum and it tries to make all operations to take constant time, so there is no easy way to get the value out of it

Reply:
I took a look at the [`CtOption` doc page](https://docs.rs/subtle/2.1.1/subtle/struct.CtOption.html) and it specifically mentions finite fields, so yeah

Reply:
it implements `From<CtOption<T>>` for `Option<T>` but I am not sure it is the right thing to do because that will likely violate the constness of this API

Reply:
We talked about it offline and the ideal solution would be to convert `CtOption` to `Result<T, E>` but this change has to be made in `CtOption` itself and this struct lives in the external crate.

Reply:
You mean something like `Option::ok_or` ?

Reply:
yea, that was the method I was thinking about

Reply:
resolved?